RPD refactor

Citinited opened the pull request:
This PR refactors the RPD. I've been meaning to do this for a while, as while the code works, it was my first large project and therefore could be improved upon greatly.

Summary of changes:
- Instead of a huge `attackby` proc, the RPD now calls `rpd_act`. This simplifies things, and keeps things more sane.
- Refactors and cleans up `rpd.tmpl` so that it doesn't evaluate the same thing multiple times, and looks neater.
- Removes `RPD_ALLOWED_HERE` flag as it is now redundant (this is instead handled by `rpd_act`).
- Moves pipe defines to their own file, and keeps them as global defines (i.e. they are not undefined anywhere).
- Refactors `pipelist` and `dpipelist` vars. Instead of storing pipe data in a stupidly large list of pipes, it is stored as `/datum/pipes`. The two lists have also been consolidated into one.
- Refactors `/obj/structure/disposalconstruct/New` so that it actually makes the pipe you want to make without having to add code afterwards to do so.
- Refactors disposals pipe dispenser to make that stupid `switch(p_type)` proc better.
- Removes 8 white pixels from the disposals pipe junction (these served no purpose, I think the original spriter just forgot to remove them)

:cl:
tweak: RPD behaviour has been slightly tweaked - for example you can now rotate / flip / delete individual pipes by clicking on them without affecting other pipes on that tile.
/:cl:

Citinited commented on code/__HELPERS/global_lists.dm:
```diff
@@ -50,6 +50,13 @@
+	init_subtypes(/datum/pipes/disposal, pipe_list)
+	for(var/I in pipe_list)
+		var/datum/pipes/P = I
+		construction_pipe_list += list(list("pipename" = P.pipename, "pipeid" = P.pipeid, "atmosordisposals" = P.pipetype, "category" = P.category, "orientations" = P.orientations, "icon" = P.previewicon, "bendy" = P.bendy))
```
Can nanoUI read data directly from datums, and if so, how? That might be better than converting the datum to a list.

Tayyyyyyy commented on code/__HELPERS/global_lists.dm:
```diff
@@ -50,6 +50,13 @@
 
+//RPD stuff
+	var/list/pipe_list = list()
+	init_subtypes(/datum/pipes/atmospheric, pipe_list)
```
You can just do `var/list/pipe_list = init_subtypes(/datum/pipes/atmospheric)`!

Citinited replied:
Oh, I meant combining it with the line below i.e.

Instead of ```init_subtypes(/datum/pipes/atmospheric, pipe_list)
init_subtypes(/datum/pipes/disposal, pipe_list)```

Do something like `init_subtypes(/datum/pipes/disposal, /datum/pipes/atmospheric, pipe_list)`

Citinited commented:
I need to review this.

tigercat2000 commented on code/_globalvars/lists/objects.dm:
```diff
@@ -44,4 +44,6 @@ var/global/list/meteor_list = list()				//list of all meteors
\ No newline at end of file
+var/global/list/active_diseases = list() 			//List of Active disease in all mobs; purely for quick referencing.
+
+var/global/list/construction_pipe_list = list()		//List of all pipe datums
```
@Citinited Global lists are literally just a `var/list/x` on the `GLOB` object, you interact with them like any other list. Just make this `GLOBAL_LIST_EMPTY(construction_pipe_list)` and add stuff to it like `GLOB.construction_pipe_list += x`

Citinited replied:
Gotcha, thanks

Citinited commented:
Step 1: GLOBify stuff - complete
Step 2: nanoUI stuff - working on it

Citinited commented:
Okay, I changed my mind about `construction_pipe_list` containing datums, and I decided to leave it as a list of lists. I also cleaned up the nanoUI and some other files.

I think this is ready for reviewing.

Citinited commented on code/game/objects/items/weapons/rpd.dm:
```diff
@@ -43,96 +23,113 @@
-	list("pipename" = "Bin", "id" = 6, "orientations" = 1, "icon" = "condisposal"),
-	list("pipename" = "Outlet", "id" = 7, "orientations" = 4, "icon" = "outlet"),
-	list("pipename" = "Chute", "id" = 8, "orientations" = 4, "icon" = "intake"))
 var/list/mainmenu = list(
```
There must be a better way to do these two lists, but I'm not sure what 🤔

Citinited commented on code/game/objects/items/weapons/rpd.dm:
```diff
@@ -43,96 +23,126 @@
+
+/obj/item/rpd/proc/create_atmos_pipe(mob/user, turf/T) //Make an atmos pipe, meter, or gas sensor
+	if(!can_dispense_pipe(whatpipe, RPD_ATMOS_MODE))
+		to_chat(user, "<span class='warning'>[src] beeps, \"ERROR\" </span>") //Damn dirty apes -- I mean hackers
```
@tigercat2000 Do you want this to message admins in the event that this happens?

Crazylemon64 replied:
`log_runtime(EXCEPTION("Failed to spawn pipe [whatpipe] - possible tampering detected"))` can work
Not `whatpipe` directly, of course, since that's just an ID that doesn't provide much info on its own

Crazylemon64 commented:
My apologies for taking so long to get to this one! I'll get through a review, now

Crazylemon64 commented on code/game/objects/items/weapons/rpd.dm:
```diff
@@ -43,96 +23,126 @@
 	spark_system.set_up(1, 0, src)
 	spark_system.attach(src)
 
+/obj/item/rcd/Destroy()
```
rpd/Destroy

Crazylemon64 commented on nano/templates/rpd.tmpl:
```diff
@@ -4,101 +4,104 @@ Used In File(s): /code/game/objects/items/weapons/rpd.dm
 	</div>
 	{{for data.pipelist}} <!-- For the selected pipe, we want to show a preview to allow the user to select a rotation-->
+		{{if value.pipe_type != 1 || value.pipe_id != data.whatpipe || value.orientations == 1}}	<!--Skip all this icon nonsense if the pipe is: not an atmos pipe; not the pipe we selected; or pipe has one orientation-->
+			{{continue;}}
```
I don't think this is a valid statement within `doT.js`

Crazylemon64 replied:
Nevermind, seems like it's covered in the `evaluate` clause

DesolateG commented:
Volume pumps is missing
